libs/auth/react/src/lib/Authenticator/CheckIdpSession.tsx

```diff
@@ -30,58 +22,108 @@ export interface SessionInfoResponse {
+  }, [])
+
+  const resetUserSession = useCallback(() => {
+    if (userSession.current.intervalHandle) {
+      clearInterval(userSession.current.intervalHandle)
+    }
+    userSession.current.intervalHandle = null
+    userSession.current.hasMessage = false
+    userSession.current.retryCount = 0
+    // Intentionally not resetting sessionEnd as it
+    // indicates that the user has had session before.
+  }, [])
 
   const signInRedirect = useCallback(async () => {
     await userManager.removeUser()
     return window.location.reload()
   }, [userManager])
 
-  const onLoadHandler = (event: SyntheticEvent<HTMLIFrameElement>) => {
-    if (authSettings.authority) {
-      event?.currentTarget?.contentWindow?.postMessage(
-        {
-          type: SessionInfoMessageType.SessionInfoRequest,
-        } as SessionInfoRequest,
-        authSettings.authority,
-      )
-    }
-  }
+  const checkActiveSession = useCallback(() => {
+    setTimeout(() => {
+      const hasBeenActive = !!userSession.current.sessionEnd
+      const { retryCount, hasMessage } = userSession.current
+
+      if (!isActive() && retryCount > MAX_RETRIES && hasBeenActive) {
+        window.location.reload()
+      } else if (!hasMessage && retryCount < MAX_RETRIES) {
+        userSession.current.retryCount += 1
+        reloadIframe()
+      }
+    }, ACTIVE_SESSION_DELAY)
+  }, [isActive])
 
   const messageHandler = useCallback(
     async ({ data, origin }: MessageEvent): Promise<void> => {
-      const sessionInfo = data as SessionInfoResponse
+      const sessionInfo = data as SessionInfo
 
+      // Check if the postMessage is meant for us
       if (
         origin !== authSettings.authority ||
-        sessionInfo.type !== SessionInfoMessageType.SessionInfoResponse
+        sessionInfo.type !== SessionInfoMessageType
       ) {
         return
       }
 
       if (sessionInfo && sessionInfo.message === 'OK') {
+        userSession.current.hasMessage = true
+
         // SessionInfo was found, check if it is valid or expired
         if (sessionInfo.isExpired) {
           return signInRedirect()
-        } else if (sessionInfo.expiresIn !== undefined) {
-          if (!sessionTimeout.current) {
-            const newSessionTimeout = setTimeout(() => {
-              sessionTimeout.current = null
-              setIframeChecksum((i) => i + 1)
-            }, sessionInfo.expiresIn * 1000)
-            sessionTimeout.current = newSessionTimeout
-          }
+        } else if (!isActive() && sessionInfo.expiresIn !== undefined) {
+          userSession.current.sessionEnd = addSeconds(
+            new Date(),
+            sessionInfo.expiresIn,
+          )
+
+          userSession.current.intervalHandle = setInterval(() => {
+            const now = new Date()
+
+            if (
+              userSession.current.sessionEnd &&
+              now > userSession.current.sessionEnd
+            ) {
+              resetUserSession()
+              reloadIframe()
```

Comment:
Maybe explain what this condition means. Why we reload the iframe instead of signing in.